controllers/cloudstackfailuredomain_controller.go

```diff
@@ -0,0 +1,131 @@
+	r.IsoNet = &infrav1.CloudStackIsolatedNetwork{}
+	// Setup the base runner. Initializes pointers and links reconciliation methods.
+	r.ReconciliationRunner = csCtrlrUtils.NewRunner(r, r.ReconciliationSubject, "CloudStackFailureDomain")
+
+	return r
+}
+
+// Reconcile is the method k8s will call upon a reconciliation request.
+func (reconciler *CloudStackFailureDomainReconciler) Reconcile(ctx context.Context, req ctrl.Request) (retRes ctrl.Result, retErr error) {
+	return NewCSFailureDomainReconciliationRunner().
+		UsingBaseReconciler(reconciler.ReconcilerBase).
+		ForRequest(req).
+		WithRequestCtx(ctx).
+		RunBaseReconciliationStages()
+}
+
+// Reconcile on the ReconciliationRunner actually attempts to modify or create the reconciliation subject.
+func (r *CloudStackFailureDomainReconciliationRunner) Reconcile() (retRes ctrl.Result, retErr error) {
+	res, err := r.AsFailureDomainUser(&r.ReconciliationSubject.Spec)()
+	if r.ShouldReturn(res, err) {
+		return res, err
+	}
+	// Prevent premature deletion.
+	controllerutil.AddFinalizer(r.ReconciliationSubject, infrav1.FailureDomainFinalizer)
+
+	// Start by purely data fetching information about the zone and specified network.
+	if err := r.CSUser.ResolveZone(&r.ReconciliationSubject.Spec.Zone); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "resolving CloudStack zone information")
+	}
+	if err := r.CSUser.ResolveNetworkForZone(&r.ReconciliationSubject.Spec.Zone); err != nil &&
+		!csCtrlrUtils.ContainsNoMatchSubstring(err) {
+		return ctrl.Result{}, errors.Wrap(err, "resolving Cloudstack network information")
+	}
+
+	// Check if the passed network was an isolated network or the network was missing. In either case, create a
+	// CloudStackIsolatedNetwork to manage the many intricacies and wait until CloudStackIsolatedNetwork is ready.
+	if r.ReconciliationSubject.Spec.Zone.Network.ID == "" ||
+		r.ReconciliationSubject.Spec.Zone.Network.Type == infrav1.NetworkTypeIsolated {
+		netName := r.ReconciliationSubject.Spec.Zone.Network.Name
+		if res, err := r.GenerateIsolatedNetwork(netName)(); r.ShouldReturn(res, err) {
+			return res, err
+		} else if res, err := r.GetObjectByName(r.IsoNetMetaName(netName), r.IsoNet)(); r.ShouldReturn(res, err) {
+			return res, err
+		}
+		if r.IsoNet.Name == "" {
+			return r.RequeueWithMessage("Couldn't find isolated network.")
+		}
+		if !r.IsoNet.Status.Ready {
+			return r.RequeueWithMessage("Isolated network dependency not ready.")
+		}
+	}
+	r.ReconciliationSubject.Status.Ready = true
+	return ctrl.Result{}, nil
+}
+
+// ReconcileDelete on the ReconciliationRunner attempts to delete the reconciliation subject.
+func (r *CloudStackFailureDomainReconciliationRunner) ReconcileDelete() (retRes ctrl.Result, retErr error) {
+	r.Log.Info("Deleting CloudStackFailureDomain")
+	// Address Isolated Networks.
+	if r.ReconciliationSubject.Spec.Zone.Network.Type == infrav1.NetworkTypeIsolated {
```

Comment:
Would this delete a pre-existing user-specified isolated network?

Reply:
The deletion is guarded by the CAPC creted tag. Basically if the network isn't tagged as capc-created, then it won't be deleted.

It also won't be deleted if there are any other in use tags on it.